addon/components/tri-state.js

```diff
@@ -140,142 +153,27 @@ export default Component.extend({
-    } else {
-      this.send('reloadData');
-    }
-  },
-
-  /**
-   * Call the action to flush data
-   */
-  _flushData() {
-    this.send('flushData');
-  },
-
-  /**
-   * Task that makes the request(s) provided by the `dataActions` attribute. We wrap the
-   * request in a concurrency task so we have more control over the state of the request(s).
-   * @return {Promise}
-   */
-  _fetchDataTask: task(function* (actions) {
-    try {
-      // Handle the request(s) differently depending on how the actions are provided
-      switch (typeOf(actions)) {
-        case 'object': {
-          const promises = {};
-
-          // Construct an object containing promises for each action
-          for (let i = 0, keys = Object.keys(actions); i < keys.length; i += 1) {
-            promises[keys[i]] = actions[keys[i]]();
-          }
-
-          if (this.get('resolveAll')) {
-            return yield RSVP.hashSettled(promises);
-          }
-
-          return yield RSVP.hash(promises);
-        }
-
-        case 'array': {
-          const promises = actions.map((action) => {
-            return action();
-          });
-
-          if (this.get('resolveAll')) {
-            return yield RSVP.allSettled(promises);
-          }
-
-          return yield RSVP.all(promises);
-        }
-
-        default:
-          return yield actions();
+    // If there are no promises to act upon fallback to last resolved data or exit with no state
+    if (isEmpty(this.promises)) {
+      if (isPresent(this._lastResolvedData) && (this.showLastSuccessful === true)) {
+        this._setComponentState(STATE.RESOLVED);
       }
-    } catch (e) {
-      throw e;
+      return this._setComponentState(STATE.NONE);
     }
-  }).cancelOn('willDestroyElement').restartable(),
-
-  actions: {
-    /**
-     * Taskify the actions provided via `dataActions` and fetch the data
-     * @return {Promise}
-     */
-    fetchData(actions) {
-      let promiseActions = actions;
-
-      if (isBlank(promiseActions)) {
-        Logger.warn([
-          'No actions were provided to the "dataActions" attribute of the tri-state component.',
-          'Provide one or more actions that each return a promise.',
-        ].join(' '));
-
-        return Promise.resolve(undefined);
-      }
 
-      // Keep a reference to the most recent request so we can reload it
-      this.set('dataActions', promiseActions);
+    // If there are some promises assume loading state
+    this._setComponentState(STATE.LOADING);
 
-      // Create new object to avoid bug with mutating the hash helper object
-      if (typeOf(promiseActions) === 'object') {
-        promiseActions = merge({}, promiseActions);
+    // Once the promises settle then cache resolved data and set the final state
+    this._computePromiseState().then(([promiseState, data]) => {
+      if (promiseState === STATE.RESOLVED) {
+        this._lastResolvedData = data;
+        this.sendAction('onResolvedData', data);
```

Comment:
Not a fan of bubbling actions. We should require a callback to be passed in to handle this "event".